### PR TITLE
Interactive Beam -- pipeline analyzer

### PR DESCRIPTION
**Pull request by qinyeli**
Main changes:
* Moved pipeline analysis out of interactive_runner and added unit tests

Aim of this commit is to
1. Increase test coverage.
2. Make it easier for possible suport for other languages in the future.

**Please** add a meaningful description for your change here

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

File: sdks/python/apache_beam/runners/interactive/pipeline_analyzer.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        return_context=True)
+
+    # context returned from to_runner_api is more informative than that returned
+    # from from_runner_api.
+    _, context = pipeline.to_runner_api(return_context=True)
+
+    pipeline_info = PipelineInfo(self._pipeline_proto.components)
+    self._caches_used = set()
+
+    def _producing_transforms(pcoll_id, leaf=False):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/interactive/pipeline_analyzer.py]: Does this need to be a nested function?  Can we move this to directly in the class and give it better documentation?

File: sdks/python/apache_beam/runners/interactive/pipeline_analyzer.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    self._underlying_runner = underlying_runner
+    self._desired_cache_labels = desired_cache_labels or []
+
+    self._pipeline_proto_to_execute = None
+    self._top_level_referenced_pcollection_ids = None
+    self._top_level_required_transforms = None
+
+    self._analyze_pipeline()
+
+  def _analyze_pipeline(self):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/interactive/pipeline_analyzer.py]: Can you make this method smaller by factoring out logical subroutines or by adding comments for the intent of each block?  Right now, it is hard to understand the intent sometimes without reading through the entire file.  For example, what "desired" and "Derivation" here are not very clearly defined or documented (or in the corresponding methods / classes), so it's not very easy to understand how we intend for these to be used.

**Comment by qinyeli** [comment]
.

**Comment by charlesccychen** [comment]
Thanks, this LGTM.

**Comment by qinyeli** [comment]
As a note:
This PR resolved BEAM-4755 as it preserves the composite structure of PTransforms.

**Pull request merged**